src/dos/program_more.cpp

```diff
@@ -0,0 +1,335 @@
+#include "checks.h"
+#include "string_utils.h"
+
+#include <algorithm>
+#include <cctype>
+#include <vector>
+
+CHECK_NARROWING();
+
+
+// ASCII control characters
+static constexpr char code_ctrl_c    = 0x03; // end of text
+static constexpr char code_lf        = 0x0a; // line feed
+static constexpr char code_cr        = 0x0d; // carriage return
+static constexpr char code_esc       = 0x1b; // escape
+
+
+void MORE::Run()
+{
+	// Handle command line
+	if (HelpRequested()) {
+		WriteOut(MSG_Get("SHELL_CMD_MORE_HELP_LONG"));
+		return;
+	}
+	if (!ParseCommandLine())
+		return;
+
+	// Check screen height
+	constexpr uint16_t min = 10;
+	new_line_counter_start = std::max(min, INT10_GetTextRows());
+	new_line_counter_start = static_cast<uint16_t>(new_line_counter_start - 2);
+
+	if (mode_singlefile) {
+		// XXX
+	} else if (mode_multifile) {
+		// XXX
+	} else {
+		// We need to be able to read STDIN for key presses, but it is most
+		// likely redirected - so clone the handle, and reconstruct real STDIN
+		// from STDERR (idea from FreeDOS implementation,
+		// https://github.com/FDOS/more/blob/master/src/more.c)
+
+		if (!DOS_DuplicateEntry(STDIN, &input_handle) ||
+		    !DOS_ForceDuplicateEntry(STDERR, STDIN)) {
+			LOG_ERR("DOS: Unable to prepare handles in MORE.COM");
+			return;
+		}
+
+		ProcessInputStream();
+	}
+
+	WriteOut("\n");
+}
+
+bool MORE::ParseCommandLine()
+{
+	// Put all the parameters into vector, reverse their order
+	cmd->FillVector(params);
+
+	// Check if specified tabulation size
+	if (!params.empty()) {
+		const auto &param = params[0];
+		if ((starts_with("/t", param) || starts_with("/T", param)) &&
+			(param.length() == 3) && (param.back() >= '1') && (param.back() <= '9'))
+		{
+			// FreeDOS extension - tab size
+			tab_size = static_cast<uint8_t>(param.back() - '0');
+			params.erase(params.begin());
+		}
+	}
+
+	// Make sure no other switches are supplied
+	for (const auto &param : params)
+		if (starts_with("/", param)) {
+			WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH"), param.c_str());
+			return false;
+		}
+
+	// Determine if single/multi file processing
+	if (params.size() > 1)
+		mode_multifile = true;
+	else if (params.size() == 1) {
+		if (contains(params[0], '*') || contains(params[0], '?'))
+			mode_multifile = true;
+		else
+			mode_singlefile = true;
+	}
+
+	return true;
+}
+
+MORE::Decision MORE::ProcessInputStream()
+{
+    bool     new_line         = false;
+	uint16_t new_line_counter = 0;
+	auto     previous_column  = GetCurrentColumn();
+
+	skip_cr = false;
+	skip_lf = false;
+
+    Decision decision = Decision::NextFile;
+	while (true) {
+		if (shutdown_requested) {
+			decision = Decision::Terminate;
+			break;
+		}
+
+		// Read character
+		char code = 0;
+		if (!GetCharacter(code)) {
+			decision = Decision::NextFile; // end of file
+			break;
+		}
+
+		// Reset new line counter if needed
+		if (!new_line_counter || !GetCurrentRow())
+			new_line_counter = new_line_counter_start;
```

Comment:
We're relying on `new_line_counter_start` being > 0 here; because if it's 0, then `new_line_counter = 0` and further down `--new_line_counter;` will wrap to UINT16_MAX.

So suggest asserting that `new_line_counter_start > 0` here (or as a pre-condition before the while loop.. ) Just a bit of assumption checking.

Reply:
This was changed, line counter now starts from 0 and goes upwards - as this was more convenient for the improvements I made.